src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -0,0 +1,78 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/ast_construction.h"
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "src/common/testing/gtest.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+TEST(AstConstructionSuite, SimpleQueryTest) {
+  absl::string_view input_text =
+      R"qql(may_notifierA = query "UserA" says mayA("NotifierA")
+                                         may_notifierB = query "UserB" says mayB("NotifierB"))qql";
+
+  Program prog_out = ParseProgram(input_text);
+  std::vector<std::string> actual_name;
+  std::vector<std::string> actual_principal_name;
+  std::vector<std::string> actual_predicate_name;
+  std::vector<std::string> actual_predicate_args;
+
+  for (auto const &each_query : prog_out.queries()) {
+    actual_name.push_back(each_query.name());
+    actual_principal_name.push_back(each_query.principal().name());
+    BaseFact basefact(
+        std::move(std::get<BaseFact>(each_query.fact().GetValue())));
+    actual_predicate_name.push_back(
+        std::get<datalog::Predicate>(basefact.GetValue()).name());
+    for (auto const &each_arg :
+         std::get<datalog::Predicate>(basefact.GetValue()).args()) {
+      actual_predicate_args.push_back(each_arg);
+    }
+  }
+
+  std::string expected_name = R"(may_notifierA)";
+  std::string another_expected_name = R"(may_notifierB)";
+  std::string expected_principal_name = R"("UserA")";
+  std::string another_expected_principal_name = R"("UserB")";
+  std::string expected_predicate_name = R"(mayA)";
+  std::string another_expected_predicate_name = R"(mayB)";
+  std::string expected_predicate_args = R"("NotifierA")";
+  std::string another_expected_predicate_args = R"("NotifierB")";
+
+  EXPECT_EQ(actual_name.at(0), expected_name);
+  EXPECT_EQ(actual_name.at(1), another_expected_name);
+  EXPECT_EQ(actual_principal_name.at(0), expected_principal_name);
+  EXPECT_EQ(actual_principal_name.at(1), another_expected_principal_name);
+  EXPECT_EQ(actual_predicate_name.at(0), expected_predicate_name);
+  EXPECT_EQ(actual_predicate_name.at(1), another_expected_predicate_name);
+  EXPECT_EQ(actual_predicate_args.at(0), expected_predicate_args);
+  EXPECT_EQ(actual_predicate_args.at(1), another_expected_predicate_args);
```

Comment:
User matchers to make the test more readable:

```suggestion
Predicate mayA("mayA", {"NotiferA"}, datalogg::Sign::kPositive);
... 
EXPECT_THAT(prog_out.queries(), 
  testing::UnorderedElementsAre(
     Query("may_notifierA", Principal("UserA"), Fact(BaseFact(mayA))),
     Query(...)
```

If you do care about the order user `testing::ElementsAre` instead of `testing::UnorderedElementsAre`.